src/back/spv/image.rs

```diff
@@ -314,41 +314,76 @@ impl<'w> BlockContext<'w> {
+                kind,
+                crate::ScalarKind::Sint | crate::ScalarKind::Uint
             ));
-            reconciled_id
+            kind
         } else {
-            let component_type_id = self.get_type_id(LookupType::Local(LocalType::Value {
-                vector_size: None,
-                kind: component_kind,
-                width: 4,
-                pointer_space: None,
-            }));
-
-            let reconciled_id = self.gen_id();
-            block.body.push(Instruction::unary(
-                spirv::Op::ConvertUToF,
-                component_type_id,
-                reconciled_id,
-                array_index_i32_id,
-            ));
-            reconciled_id
+            unreachable!("we only allow i32 and u32");
+        };
+        let reconciled_array_index_id = match (component_kind, array_index_kind) {
+            (crate::ScalarKind::Sint, crate::ScalarKind::Sint)
+            | (crate::ScalarKind::Uint, crate::ScalarKind::Uint) => array_index_id,
+            (crate::ScalarKind::Sint, crate::ScalarKind::Uint) => {
+                let i32_id = self.get_type_id(LookupType::Local(LocalType::Value {
+                    vector_size: None,
+                    kind: crate::ScalarKind::Sint,
+                    width: 4,
+                    pointer_space: None,
+                }));
+
+                let reconciled_id = self.gen_id();
+                block.body.push(Instruction::unary(
+                    spirv::Op::Bitcast,
+                    i32_id,
+                    reconciled_id,
+                    array_index_id,
+                ));
+                reconciled_id
+            }
+            (crate::ScalarKind::Uint, crate::ScalarKind::Sint) => {
+                let u32_id = self.get_type_id(LookupType::Local(LocalType::Value {
+                    vector_size: None,
+                    kind: crate::ScalarKind::Uint,
+                    width: 4,
+                    pointer_space: None,
+                }));
+
+                let reconciled_id = self.gen_id();
+                block.body.push(Instruction::unary(
+                    spirv::Op::Bitcast,
+                    u32_id,
+                    reconciled_id,
+                    array_index_id,
+                ));
+                reconciled_id
+            }
+            (crate::ScalarKind::Float, _) => {
+                let component_type_id = self.get_type_id(LookupType::Local(LocalType::Value {
+                    vector_size: None,
+                    kind: component_kind,
+                    width: 4,
+                    pointer_space: None,
+                }));
+
+                let reconciled_id = self.gen_id();
+                block.body.push(Instruction::unary(
+                    spirv::Op::ConvertUToF,
+                    component_type_id,
+                    reconciled_id,
+                    array_index_id,
+                ));
+                reconciled_id
+            }
```

Comment:
I think we can reduce the repetition in this `match` by returning an `Option` with the spirv operation and writing the conversion after.

Reply:
Done, much better indeed.